### PR TITLE
W&B Tables to view sample of retrofit examples over time.

### PR DESCRIPTION
**Pull request by jscuds**
## Thus far:

- [x] added `elmo_sentence_decode` and `elmo_word_decode` to `utils.py`; allows you to turn an ELMo sentence tensor into a readable string
- [x] made `wb_table: TableLog` an attribute of `RetrofitExperiment`, 
- [x] modified `retrofit_hinge_loss` to update internal dictionary attributes with the hinge loss, pos_word_dist, neg_word_dist; 
- [x] make sure the table lists the positive/negative example index (for grouping pos/neg examples in a given epoch/split)
- [x]  loop over final dictionaries at end of training--> `update_wandb_table()`
- [x] ensured the table logs each epoch; total number of rows should be `num_table_examples` * 4 * `epochs` (4 because pos/neg for each example and train/val examples)
- [x] addressed time complexity issue
- [x] added test coverage for `TableLog`
- [ ] address the "minor improvement"

### CODE REVIEW

**Comment by jxmorris12** [comment]
closes #17

File: utils.py [hunk 1]
```diff
@@ -53,3 +53,37 @@ def log_wandb_histogram(list_of_values: list, description: str, step: int, epoch
...
+    sent = tuple([tuple(word.tolist()) for word in sent.squeeze()])
+    full_sent = []
+    for word in sent:
+        if word == tuple([0]*50): continue
+        word_concat = []
+        for char in word:
+            # print(char)
+            if char == 259: continue
+            if char == 260: break
+            if char == 261: break
```
Comment (jxmorris12) [inline utils.py]: what are these? (and how do you know?)

File: utils.py [hunk 2]
```diff
@@ -53,3 +53,37 @@ def log_wandb_histogram(list_of_values: list, description: str, step: int, epoch
...
+
+def elmo_word_decode(word: Tuple) -> str:
+    """
+    Decodes a word tuple of ELMo character_ids with length=50 into a readable sentence string.
+    Useful for debugging when you only have a ParaDatasetElmo object and its various tensor/tuple/sent_id returns.
+    """
+    word_concat = []
+    for char in word:
+        if char == 259: continue
+        if char == 260: break
```
Comment (jxmorris12) [inline utils.py]: same question as above, please add explanatory comment

File: experiment.py [hunk 3]
```diff
@@ -182,6 +200,17 @@ def retrofit_hinge_loss(self,
...
+        # Populate tracked examples in dictionaries for W&B table.
+        # Handle exception for test_loss.py because RetrofitExperiment.get_dataloaders() isn't called.
+        try:
+            if len(self.wb_table.batch_indices) > 0:
+                for ex_idx, b_idx in self.wb_table.batch_indices.items():
+                    self.wb_table.table_hinge_loss[ex_idx] = loss[b_idx]
+                    self.wb_table.table_pos_word_dist[ex_idx] = positive_pair_distance[b_idx]
+                    self.wb_table.table_neg_word_dist[ex_idx] = negative_pair_distance[b_idx]
+        except AttributeError:
+            pass
```
Comment (jxmorris12) [inline experiment.py]: Instead of using `try-except`, you should check set `self.wb_table` to None in the init, and here run this code `if self.wb_table is not None`.

File: experiment.py [hunk 4]
```diff
@@ -167,8 +185,8 @@ def retrofit_hinge_loss(self,
         """
         assert word_rep_pos_1.shape == word_rep_pos_2.shape
         assert word_rep_neg_1.shape == word_rep_neg_2.shape
-        positive_pair_distance = torch.norm(word_rep_pos_1 - word_rep_pos_2, p=2, dim=1) # TODO: need keepdim=True??
-        negative_pair_distance = torch.norm(word_rep_neg_1 - word_rep_neg_2, p=2, dim=1) # shape: (batch_size,) if keepdim=False
+        positive_pair_distance = torch.norm(word_rep_pos_1 - word_rep_pos_2, p=2, dim=-1) # NOTE(js) changed from dim=1 to dim=-1 to work with batch_size=1
```
Comment (jxmorris12) [inline experiment.py]: I don't think this is a good idea! Instead of changing the loss function here, can you change the test to add a fake batch dimension?

File: tablelog.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            self.table_pos_sent1[i] = pos_sent1
+            self.table_pos_sent2[i] = pos_sent2
+            self.table_pos_sent1_str[i] = pos_sent1_str
+            self.table_pos_sent2_str[i] = pos_sent2_str
+            self.table_pos_word1[i] = token1
+            self.table_pos_word2[i] = token2
+            assert pos_sent1_str.split(' ')[token1] == pos_sent2_str.split(' ')[token2] # assert that the word is the same in both pos_sent1 & pos_sent2
+            self.table_pos_target_word[i] = pos_sent1_str.split(' ')[token1]
+
+    # TODO: called every time a batch is passed in the train/val loop from the respective dataloader
```
Comment (jxmorris12) [inline tablelog.py]: why is this a TODO?

File: tablelog.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import time
+from pathlib import Path
+from typing import Dict, List, Tuple
+
+import numpy as np
+import pytest
+import torch
+import tqdm
+import wandb
+from torch.utils.data import DataLoader
```
Comment (jxmorris12) [inline tablelog.py]: can you change this to only import what you need, I don't think you're using some of this stuff (definitely not argparse or shutil or pytest)

File: utils.py [hunk 7]
```diff
@@ -53,3 +53,37 @@ def log_wandb_histogram(list_of_values: list, description: str, step: int, epoch
...
+def elmo_sentence_decode(sent: torch.Tensor) -> str:
+    """
+    Decodes a sentence of ELMo character_ids with the shape [1,sequence_length,50] into a readable sentence string.
+    Useful for debugging when you only have a ParaDatasetElmo object and its various tensor/tuple/sent_id returns.
+    """
+    sent = tuple([tuple(word.tolist()) for word in sent.squeeze()])
+    full_sent = []
+    for word in sent:
+        if word == tuple([0]*50): continue
+        word_concat = []
```
Comment (jxmorris12) [inline utils.py]: nit: can you call this `chars_in_word`?

**Review by jxmorris12 — COMMENTED** [review]
Thanks for solving this, this was a hard engineering problem, and the code overall looks good! I left a few minor comments, and I have two big comments. Let me know if you have questions, especially about the second thing.

### Tests

TableLog needs test coverage. here are a couple options of how to test:
- Directly add things to the table and then check the contents of the table somehow `assert self.final_table.data == x`, or something like that, I'm not actually sure what the API for `self.final_table` looks like
- You could try to do some sort of fake experiment and check the table at the end of the experiment too

### Looping over the dataset

It seems like you're looping over the full dataset. This is a big problem and will probably make retrofitting a lot slower. I think this happens because `experiment.wb_table.check_tracked_indices` is called on every single training and testing step, and that function contains a loop over all of the examples in the batch! I think we never want to loop over every single thing in the full training set (even in the main training loop, we iterate over *batches*, not *examples*).

There are various ways to fix this:
1. Give up on the idea of logging the same examples every time. Just log a new table with new examples on every epoch. Like get the first examples or the last ones or something and just use those. 
2. Pass in a fixed batch at the beginning and re-get the loss for it at logging time. This is probably what I would do. I think it's fine to recompute the loss for a single extra batch during evaluation, especially as opposed to looping over the full training set every single time.

## A minor improvement

Not sure how easy this is, but it could be cool to type out the strings from the table in the paper and add them to the table every time.  That way part of our table would be exactly like the one in the paper. (Don't worry about it if that sounds too hard or just not interesting)

File: tablelog.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+
+from utils import elmo_sentence_decode
+
+class TableLog():
```
Comment (jxmorris12) [inline tablelog.py]: oh no need for empty parentheses here, can just do `class TableLog:`

File: tablelog.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    """A class to log W&B tables for specific examples over each epoch of retrofitting."""
+    
+    def __init__(self, num_table_examples:int, train_dataloader: DataLoader, test_dataloader: DataLoader, columns:List[str]):
```
Comment (jxmorris12) [inline tablelog.py]: nit: put a space between colon and type, like `num_table_examples: int` not `num_table_examples:int`

File: utils.py (hunk 1 above)
Comment (jscuds) [reply utils.py]: **Short Answer**:
- 259 is the beginning of word character
- 260 is the end of word character
- 261 is the padding character

**More Detail**:
From [AllenNLP's ELMo Tokenizer](https://github.com/allenai/allennlp/blob/b8f92f036e37015aabe0d76aabf2722597c7686f/allennlp/data/token_indexers/elmo_indexer.py#L38-L44) "char ids 0-255 come from utf-8 encoding bytes."  Based on trial and error, I realized that for some reason, the values are off by 1.  Example: 260 should be the "padding character", but in reality it's 261.  I have no idea why this is the case...but my decoder works empirically.

File: experiment.py (hunk 4 above)
Comment (jscuds) [reply experiment.py]: I didn't have to adjust any tests, but I'll change it back.

File: tablelog.py (hunk 5 above)
Comment (jscuds) [reply tablelog.py]: Left accidently - removing.
